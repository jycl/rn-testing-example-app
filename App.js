/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow
 */

import React, {Component} from 'react';
import {Platform, StyleSheet, Text, View} from 'react-native';
import { Provider } from "./src/components/Context";
import Testable from "./src/components/Testable";
import Clickable from './src/components/Clickable';
import TestableList from './src/components/TestableList';
import {getUser} from "./src/services/github";

const instructions = Platform.select({
  ios: 'Press Cmd+R to reload,\n' + 'Cmd+D or shake for dev menu',
  android:
    'Double tap R on your keyboard to reload,\n' +
    'Shake or press menu button for dev menu',
});

const sampleData = [
  {name: "Joe"},
  {name: "Pete"},
  {name: "Mickey"},
  {name: "Mike"},
];

export default class App extends Component {
  state = {githubUser: {}};

  componentDidMount() {
    getUser('jycl')
    .then(json => {
      this.setState({
        githubUser:json
      })
    });
  }

  render() {
    return (
      <Provider value="red">
        <View testID="welcome" style={styles.container}>
          <Text>Testing components using react-test-renderer</Text>
          <Testable pair="blue"/>
          <Clickable testID="button" />
        </View>
        <TestableList data={sampleData}/>
        <Text style={styles.container}>{this.state.githubUser.login || "None yet."}</Text>
      </Provider>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF',
  },
  welcome: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10,
  },
  instructions: {
    textAlign: 'center',
    color: '#333333',
    marginBottom: 5,
  },
});
