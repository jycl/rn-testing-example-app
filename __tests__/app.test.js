import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import App from "../App";
import {FlatList, Text, TouchableHighlight, Image} from 'react-native';

//"react-test-renderer" library tutorials / examples (without Enzyme):
/**
 * https://medium.com/@pshrmn/testing-react-native-components-in-node-with-react-test-renderer-cb2985402dce 
 * 1. Rendering elements
 * 2. Testing props
 * 3. Finding nested elements
 * 4. Updating props
 * 5. Calling event handlers
 */

/**
 * Testing React 16.3 without Enzyme
 * https://itnext.io/testing-react-16-3-components-with-react-test-renderer-without-enzyme-d9c65d689e88 
 * TestRenderer.create() returns renderer (wrapper around components)
 *     - .toJSON() returns a JavaScript object representing HTML/RN output
 *     - .root reference to root component of component tree
 * 
 * instance (.root) methods
 *     - .find()/findAll() - takes predicate function (e.g. (el) => el.type == 'span');
 *     - .findByType()/findAllByType() - take component constructor
 *     - .findByProps/findAllByProps() - search for components by props (e.g. {name: "Joe"})
 */

 /**
  * ReactJS Test Renderer Documentation
  * https://reactjs.org/docs/test-renderer.html 
  */

//note about jest --updateSnapshot, since it is added to scripts you can use npm test -- --updateSnapshot (https://stackoverflow.com/questions/50138532/)
//see https://www.reactnative.guide/7-testing/7.1-jest-setup.html for jest setup shortcuts / scripts

describe("<App>", () => {
    it("Renders correctly with expected components", () => {
        const render = renderer.create(<App/>);
        expect(render.toJSON()).toMatchSnapshot();

        //find(predicate function with custom checking)
        //without "All", matching more than one will throw an exception
        const button = render.root.find((comp) => {
            return comp.type === TouchableHighlight
        });
        expect(button.type).toBe(TouchableHighlight);
        expect(button.props.children).toBeTruthy();

        //findAllByType get an array of all components
        const texts = render.root.findAllByType(Text);
        expect(texts).toHaveLength(8);
        texts.forEach((text) => {
            expect(text.props.children).toBeTruthy();
        });

        //use findAll to get an absence of component
        const images = render.root.findAllByType(Image);
        expect(images).toHaveLength(0);

        //finding all sub-components in a component
        const list = render.root.findByType(FlatList);
        const listItems = list.findAllByType(Text);
        expect(listItems).toHaveLength(4);

        //by props
        const component = render.root.findByProps({pair: "blue"});
        expect(component).toBeTruthy();
    });
});