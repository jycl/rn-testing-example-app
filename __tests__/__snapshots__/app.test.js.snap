// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App> Renders correctly with expected components 1`] = `
Array [
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5FCFF",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Text>
      Testing components using react-test-renderer
    </Text>
    <View>
      <Text>
        red
         & 
        blue
      </Text>
    </View>
    <View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={null}
      >
        <Text
          style={null}
        >
          1
        </Text>
      </View>
    </View>
  </View>,
  <RCTScrollView
    data={
      Array [
        Object {
          "name": "Joe",
        },
        Object {
          "name": "Pete",
        },
        Object {
          "name": "Mickey",
        },
        Object {
          "name": "Mike",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          1. Joe
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          2. Pete
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          3. Mickey
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          4. Mike
        </Text>
      </View>
    </View>
  </RCTScrollView>,
  <Text
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5FCFF",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    None yet.
  </Text>,
]
`;
